pkg/printers/internalversion/printers_test.go

```diff
@@ -1919,18 +1919,42 @@ type stringTestList []struct {
 	name, got, exp string
 }
 
-func TestTranslateTimestamp(t *testing.T) {
+func TestTranslateTimestampSince(t *testing.T) {
 	tl := stringTestList{
-		{"a while from now", translateTimestamp(metav1.Time{Time: time.Now().Add(2.1e9)}), "<invalid>"},
-		{"almost now", translateTimestamp(metav1.Time{Time: time.Now().Add(1.9e9)}), "0s"},
-		{"now", translateTimestamp(metav1.Time{Time: time.Now()}), "0s"},
-		{"unknown", translateTimestamp(metav1.Time{}), "<unknown>"},
-		{"30 seconds ago", translateTimestamp(metav1.Time{Time: time.Now().Add(-3e10)}), "30s"},
-		{"5 minutes ago", translateTimestamp(metav1.Time{Time: time.Now().Add(-3e11)}), "5m"},
-		{"an hour ago", translateTimestamp(metav1.Time{Time: time.Now().Add(-6e12)}), "1h"},
-		{"2 days ago", translateTimestamp(metav1.Time{Time: time.Now().UTC().AddDate(0, 0, -2)}), "2d"},
-		{"months ago", translateTimestamp(metav1.Time{Time: time.Now().UTC().AddDate(0, 0, -90)}), "90d"},
-		{"10 years ago", translateTimestamp(metav1.Time{Time: time.Now().UTC().AddDate(-10, 0, 0)}), "10y"},
+		{"a while from now", translateTimestampSince(metav1.Time{Time: time.Now().Add(2.1e9)}), "<invalid>"},
+		{"almost now", translateTimestampSince(metav1.Time{Time: time.Now().Add(1.9e9)}), "0s"},
+		{"now", translateTimestampSince(metav1.Time{Time: time.Now()}), "0s"},
+		{"unknown", translateTimestampSince(metav1.Time{}), "<unknown>"},
+		{"30 seconds ago", translateTimestampSince(metav1.Time{Time: time.Now().Add(-3e10)}), "30s"},
+		{"5 minutes ago", translateTimestampSince(metav1.Time{Time: time.Now().Add(-3e11)}), "5m"},
+		{"an hour ago", translateTimestampSince(metav1.Time{Time: time.Now().Add(-6e12)}), "1h"},
+		{"2 days ago", translateTimestampSince(metav1.Time{Time: time.Now().UTC().AddDate(0, 0, -2)}), "2d"},
+		{"months ago", translateTimestampSince(metav1.Time{Time: time.Now().UTC().AddDate(0, 0, -90)}), "90d"},
+		{"10 years ago", translateTimestampSince(metav1.Time{Time: time.Now().UTC().AddDate(-10, 0, 0)}), "10y"},
+	}
+	for _, test := range tl {
+		if test.got != test.exp {
+			t.Errorf("On %v, expected '%v', but got '%v'",
+				test.name, test.exp, test.got)
+		}
+	}
+}
+
+func TestTranslateTimestampUntil(t *testing.T) {
+	// Since this method compares the time with time.Now() internally,
+	// I put small buffers on comparing times to consider method call overhead.
+	// Otherwise, the output strings become shorter than expected.
```

Comment:
Something is weird in my environment. In the test, for example, `translateTimestampUntil(metav1.Time{Time: time.Now().Add(3e10)` returns `29s` instead of `30s`. I will investigate and fix it.

Reply:
I think the tolerance you mentioned is [for `0s`](https://github.com/kubernetes/apimachinery/blob/97cca6cc910b472c65890ea424e84795ef4ddfb9/pkg/util/duration/duration.go#L29). So, if the overhead of the function call of `translateTimestampUntil` is 1 microsecond, it returns `29s` because the duration is `-30 seconds + 1 microsecond`. It's not the case for `translateTimestampSince` because `+30 seconds + 1 microsecond` is still `30 seconds` after the division.

May I have your suggestion to handle the function call overhead?
I usually use time freeze library such as [freezegun](https://github.com/spulec/freezegun) in Python tests, but I don't know if we have such a thing in Go.

Reply:
Looking at `ShortHumanDuration(duration)` again, it was implemented with the assumption that the input duration should not be negative (ref #12635) and it tolerates duration (0, -1) for clock skew, i.e. tolerates 1s. It's not used to handle method call overhead here. 

Therefore it's fine to add a buffer in our test. Suggest making it more explicit by defining a variable (e.g. `const buf = 1e6`). It's also fine to have a larger buffer as long as it's less than 1s.